vignettes/miaViz.Rmd

```diff
@@ -326,6 +326,127 @@ names(metadata(dmn_se))
 plotDMNFit(dmn_se, type = "laplace")
 ```
 
+# Serial data ordination and trajectories
+
+```{r, error=FALSE, warning=FALSE, results='hide'}
+if (!requireNamespace("devtools", quietly = TRUE))
+    install.packages("devtools")
+devtools::install_github("microbiome/miaTime")
+```
+
+Principal Coordinate Analysis using Bray-Curtis dissimilarity on the
+`hitchip1006` dataset:
+
+```{r}
+library(dplyr)
+data(hitchip1006, package = "miaTime")
+tse <- hitchip1006
+tse <- transformCounts(tse, method = "relabundance")
+## Ordination with PCoA with Bray-Curtis dissimilarity
+tse <- runMDS(tse, FUN = vegan::vegdist, method = "bray",
+                  name = "PCoA_BC", exprs_values = "relabundance",
+              na.rm = TRUE)
+# Calculating Explained variance
+e <- attr(reducedDim(tse, "PCoA_BC"), "eig")
+rel_eig <- e/sum(e[e>0])
+# plot
+p1 <- plotReducedDim(tse, dimred = "PCoA_BC") +
+    labs(x = paste("PCoA 1 (", round(100 * rel_eig[[1]],1), "%", ")", sep = ""),
+         y = paste("PCoA 2 (", round(100 * rel_eig[[2]],1), "%", ")", sep = ""),
+         title="PCoA - Bray-Curtis")+
+    theme(plot.title = element_text(hjust = 0.5))
+p1
+```
+
+Retrieving information about all available trajectories:
+
+```{r}
+colData(tse)$time <- as.factor(colData(tse)$time)
+# trajectories info
+traj_info <- table(colData(tse)$subject,colData(tse)$time)
+# has at least two points
+traj_info <- traj_info[rowSums(traj_info)!=1,]
+table(rowSums(traj_info)) %>% as.data.frame() %>%
+    `colnames<-`(c("Trajectories Lengths", "Counts"))
+```
+
+Lets look at all present trajectories in the data:
+
+```{r}
+# keeping only data that has one-to-one link from one point to another
+df <- traj_info %>% as.data.frame() %>%
+    `colnames<-`(c("subject", "time", "value")) %>% 
+    filter(value ==1) %>% arrange(subject, time) 
```

Comment:
Value of 0 means no trajectory present; then it is discarded.
Value of 2 or more, means replicates are present.
For instance using the `SilvermanAGutData` dataset, some subjects (Vessels in this case) had replicates, for one time point for one subject resulting in a value of 2:
`table(colData(SilvermanAGutData)$Vessel, colData(SilvermanAGutData)$DAY_ORDER)`.
Handling such replicates is to be discussed if needed; averaging their ordination coordinates values didn't seem to me very reasonable, or randomly sampling one of them nor was convincing to me. Then, thought to discard in case emerged for the moment. Unless, a manual selection of which replicate to use or some other criterion dependent on the study or the dataset itself.

To make sure that only one trajectory connects two (successive) time points, I filtered based on the value of 1.

Reply:
Ah now I get it. This could be clarified in comments. The desired solution is very case-by-case. For an example I would just remove all individuals who have duplicate measurements from single time points, before any further analysis. Comments could say that this is done for demo purposes.

Reply:
Sure, I will update the comment.

Reply:
Yep and the code can be simplified if you discard them before any analysis.

I think something like this should do the trick:
`clean_data <- full_data %>% count(subject, time) %>% filter(n==1)`

Reply:
(or perhaps `n<=1` if you would keep singletons in as well)

Reply:
I tried `colData(tse) %>% as_tibble() %>% count(subject, time) %>% filter(n==1)` does not seem to get the desired results after testing it.
`count(subject,time)` will group data per subject and count only the time point present for that subject, and additionally resulting in a count of all as 1, not enabling any filtering.

The contingency table using the `table(subject, time)` will count sample presence of the subject for all time points available in the data. The earlier step (@ line 368) `traj_info <- traj_info[rowSums(traj_info)!=1,]` is as well trivial for discarding singletons and moving on only with actual trajectories, much smaller and ordered by subject and time for the geom_line later.

However, at the moment I am thinking if all this should be redone based as well on the sample_ids as keys, so as adding it to colData, and also adding the PC 1 and 2 to colData would make the plotting in one go. 
I will give it another try.

Reply:
The suggestion you have made in later comment, has been working and solving all problems perfectly.
I tested with `selected.subjects <- names(which(table(tse$subject)==2))` for two time points and it generalizes and works well for more than two time points `selected.subjects <- names(which(table(tse$subject)>=2))`.
Thanks.

Reply:
"adding the PC 1 and 2 to colData would make the plotting in one go." might be good in this case but in the longer term we might like to invent a more generic solution on how to combined reducedDim and colData in plottings.